### PR TITLE
feat(activate): replay profile.d PATH prepends across attachments

### PR DESCRIPTION
**Pull request by limeytexan**
## Summary

Fixes a gap in the `etc/profile.d` startup-script contract delivered in #4509: scripts could prepend to `PATH`, but `flox-activations fix-paths` re-orders `PATH` at the end of every activation and attachment to keep each activated environment's `bin`/`sbin` directories foremost, demoting any such prepend behind every layered environment.

- **New helper for profile.d scripts:** `flox_prepend_path <dir>` (defined in the interpreter's `helpers.bash`) prepends `<dir>` to `PATH` immediately and registers it against `$FLOX_ENV` in `_FLOX_ENV_PATH_PREPENDS`, a colon-separated list of `<env_dir>=<prepend_dir>` pairs, newest first. Registration is idempotent, and `<dir>` must not contain `:` or `=`. Because run-mode activations do not source package-provided profile.d scripts (#4514), the helper has no effect in run mode; this is documented in `flox-activate(1)`.
- **`fix-paths` learns `--path-prepends`:** each registered directory is replayed *ahead of the registering environment's own `bin`* at that environment's position in the layered ordering — so an environment activated later still takes precedence over an earlier environment's prepends, exactly like its `bin` dirs do. Entries whose environment is no longer in `FLOX_ENV_DIRS` are dropped, and malformed entries are ignored.
- **Replay falls out of the existing machinery:** profile.d scripts run between the activation's env snapshots, so the registration is captured in the recorded activation diff and replayed into every attaching shell *before* the trailing `fix-paths` runs. The same mechanism that demoted raw prepends now re-asserts registered ones on every attach, rc re-source, and re-layering.
- **Threaded through every `fix-paths` call site:** the bash, fish, and tcsh rc generators, the Rust-side attach export computation (`VarsFromEnvironment`/`fixed_vars_to_export`), and the interpreter's `activate` and `activate.d/zsh` scripts. The tcsh generator guards the possibly-unset variable with the existing `"empty"` sentinel idiom and unsets the sentinel after use so it cannot outlive the activation.
- `flox-activate(1)` documents the helper in the profile.d section.

## Testing

- Eight unit tests in `fix_paths.rs`: layer placement, within-layer ordering of multiple registrations, dropping entries for inactive environments, entry parsing (malformed + sentinel), idempotency of re-fixing, a dir registered by two active environments (dedups to the outermost registering environment's layer), a prepend equal to the environment's own `bin` dir, and dirs containing spaces.
- Seven new bats tests:
  - in-place activation keeps the registered dir immediately ahead of the environment's `bin`/`sbin`, parametrized across bash, fish, tcsh, and zsh — the direct regression test, since the trailing `fix-paths` in the generated rc is what demoted plain prepends. The bash and tcsh variants additionally assert on deactivate that neither a registration nor the tcsh `"empty"` sentinel outlives the activation;
  - command-mode idempotency: two profile.d scripts registering the same dir yield a single `PATH` entry (command mode has no trailing re-order to dedup);
  - layered activation asserts the full ordering `inner/bin:inner/sbin:outer/extra-bin:outer/bin:outer/sbin`;
  - a genuine attach (background activation held open on a fifo) asserts the ordering for both a command-mode and an in-place attach.
- Reproduced the old behavior for comparison: a plain `PATH="$dir:$PATH"` in a profile.d script lands demoted behind all environment dirs in the same in-place flow.
- All 77 profile-related and 51 attach tests in `activate.bats` (all four shells), all 13 `activate-mode.bats` tests, and all 41 `deactivate.bats` tests pass; 108 `flox-activations` unit tests and clippy pass.
- Pre-existing local-only failures in `shell-state-restoration.bats` and `flox-rust-sdk` floxmeta git tests reproduce identically on an unmodified checkout and are unrelated to this diff.

## Release Notes

- `etc/profile.d` startup scripts can now use the `flox_prepend_path` shell function to add directories to `PATH` in a way that is preserved as environments are activated, layered, and attached to. Directories added with a plain `PATH` assignment continue to be re-ordered behind the activated environments' `bin` directories. As with all package-provided startup scripts, this has no effect in run-mode activations.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by limeytexan** [comment]
## Domain Summary

Package `profile.d` scripts can add directories to `PATH`, but Flox re-orders `PATH` at the end of every activation and attach, demoting those script prepends behind every layered environment's bin dirs. This PR adds a first-class opt-in: scripts call `flox_prepend_path "$dir"`, which registers `<env>=<dir>` into `_FLOX_ENV_PATH_PREPENDS` (plus an immediate prepend for the current shell), and fix-paths then replays each registered dir into position immediately ahead of its registering env's own `bin`/`sbin` on every re-fix. Placement is layer-aware — a later-activated env still outranks an earlier env's prepends, so a prepend behaves as an extension of its env's layer rather than a global override; within a layer the most-recent registration is foremost, registration is idempotent, prepends for no-longer-active envs are dropped from replay, and malformed entries are ignored. State rides the existing activation snapshot/replay machinery, so attaches and rc re-sources reconstruct the same ordering, and `#[serde(default)]` keeps old recorded activations deserializing cleanly. It is correctly framed as a follow-up to PR 4509.

## Description Consistency

Consistent: the mechanism, layer-aware replay, newest-first ordering, inactive-env dropping, malformed-entry handling, snapshot propagation story, per-shell threading, the tcsh sentinel dance, and the test inventory all match the diff.

Inconsistent: none.

Missing: the run-mode exclusion — `flox_prepend_path` is inert in run mode because run mode does not source package `profile.d` scripts (per PR 4509 / issue 4514). This limitation is absent from both the PR description and the new user-facing docs.

Overclaimed: the "across attachments" framing in the title and description — neither new bats test exercises a real attach (see the Code Review attach-coverage finding, which merges this note with Lens C's evidence).

## Code Review

[I] No end-to-end attach coverage despite the "across attachments" headline — `activate.bats:2772-2798`: both new bats tests are fresh activations (in-place and nested layering); neither attaches to a running activation. The attach replay path threaded through `fixed_vars_to_export` (`attach_diff/mod.rs:426`) has no bats coverage. The file's own mkfifo + background-activation attach test (`activate.bats:3698`) is the established model and would exercise exactly this path. This is the same issue as the description's "overclaimed" attach note — the title claims a behavior the tests never demonstrate.

[I] New bats tests are bash-only, against the file's per-shell convention — `activate.bats:2778,2795`: the file parametrizes attach/profile behavior across {bash,fish,tcsh,zsh} via shared helpers; this PR changes rc generation for all four shells and tcsh gets the largest, most delicate change (the sentinel setenv/unsetenv dance), yet only bash is integration-tested. The Rust unit tests verify the text of the tcsh statements, not that tcsh executes them. This repeats the bash-only gap flagged on PR 4509 — but here the feature *is* shell-specific rc generation, so at minimum a tcsh in-place case is warranted.

[I] Silent drop on env-dir mismatch with the two-arg helper form — `helpers.bash:43-45` and `fix_paths.rs:139`: matching is exact `PathBuf` equality against `FLOX_ENV_DIRS` entries. The default `${FLOX_ENV?}` path is byte-identical by construction (`set_env_dirs.rs` builds `FLOX_ENV_DIRS` from `FLOX_ENV` verbatim), but the documented optional `[<flox_env>]` argument lets a caller register a path differing by a trailing slash or symlink, and the prepend then vanishes with no diagnostic. Either drop the second argument until needed, document the exact-match requirement, or log dropped entries.

[M] Directory paths containing `:` or `=` break the registration/replay round-trip — `helpers.bash:46-58` and `fix_paths.rs:83-95`: entries are stored as `<env>=<dir>` joined by `:` with no escaping, so a dir containing `:` is torn into two entries and a dir containing `=` mis-parses. The helper's immediate PATH prepend uses the full dir, so the in-shell PATH and the replayed PATH diverge (the prepend is demoted on the next attach). Pathological but legal paths; a documented constraint at minimum.

[M] Prepend values inherit PATH's existing quoting posture — `fix_paths.rs:46-64`: replayed dirs are emitted as `export PATH="{new_path}"` then eval'd, so a dir containing `"`, `$`, or a backtick breaks out, exactly as a malicious PATH entry already can. Same trust level as PATH, no new boundary crossed — verified to match existing behavior, not separately hardened.

[M] Version-skew coupling — `activate:198`, `activate.d/zsh:40`, and three `gen_rc` emitters now unconditionally pass `--path-prepends`: a newer interpreter paired with an older `flox-activations` binary fails with "unexpected argument", aborting activation. The reverse direction is safe (`default_value = ""`). This is the standard interpreter/binary lockstep for this repo, but worth confirming against the rollout model.

[M] `_FLOX_ENV_PATH_PREPENDS` is never pruned, only filtered at PATH-emit time — `fix_paths.rs:105-143`: stale `<env>=<dir>` pairs for deactivated envs persist in the variable indefinitely and are re-parsed on each fix-paths call. Growth is bounded in practice (per layered env x prepends). A comment noting "entries accumulate; filtering happens at replay" would help; pruning in the emitted export would be cleaner if it ever matters.

[M] Deactivate behavior unverified for the new variable — `_FLOX_ENV_PATH_PREPENDS` is not in the explicit unset list that covers `FLOX_ENV_DIRS` et al. Env-snapshot restoration plausibly removes it and stale entries self-clean at consumption, but no deactivate test asserts the var does not outlive deactivation.

[M] Unit test edge-case gaps — `fix_paths.rs:373-425`: the five new tests are well-chosen and stylistically consistent, but untested cases include the same prepend dir under two different envs, a prepend dir equal to an env's own bin dir, and dirs containing spaces or `=`.

[M] Unconditional re-prepend on a case-guard miss — `helpers.bash:46-47`: the quoted expansion correctly makes glob chars inert, but a re-registration that fails the exact match still re-prepends to PATH unconditionally, relying on the next fix-paths to dedup. Behavior is correct but non-obvious; a one-line comment is warranted.

[M] Serde attribute inconsistency — `vars_from_env.rs`: `path_prepends` uses a bare `#[serde(default)]` while the sibling `full_env` adds `skip_serializing_if`. The bare form serializes `"path_prepends": null` into recorded state — harmless, but inconsistent with the neighbor.

[M] Hot path is now O(env_dirs x prepends) — `fix_paths.rs:138-142`: both factors scale with layering depth (typically <10), so cost is negligible; no new exec of the Rust binary, it rides the existing trailing fix-paths call. On record only.

[M] tcsh diverges structurally but justifiably — `tcsh.rs:202-224`: it reuses the existing "empty" sentinel idiom (as `FLOX_ENV_DIRS`/`MANPATH` already do) because tcsh cannot pass unset/empty args, and additionally unsetenvs the sentinel after use so it cannot leak. Rust handles the sentinel in one place (`separate_prepends_list`). Right seam, established pattern.

[M] `--path-prepends` threading is duplicated across every fix-paths call site (bash/fish/tcsh `gen_rc`, `activate`, `activate.d/zsh`, `fixed_vars_to_export`) — inherent to the per-dialect design and matching how `--path`/`--manpath`/`--env-dirs` are threaded; a future 6th argument touches all sites again, and only tests would catch a miss.

Positives: the Rust changes are clean and idiomatic — `separate_prepends_list` mirrors existing naming, comments explain the reverse-iteration ordering, and doc accuracy checks out (`$FLOX_ENV` is exported before `profile.d` runs). The architecture is well-factored: shell registers, Rust replays, reusing the exact `prepend_dirs_to_pathlike_var` machinery so prepends inherit per-layer precedence, with no new process spawns on the hot path. Trust boundary is consistent on attach; there is no unbounded growth or duplicate accumulation (`profile.d` guarded by `_FLOX_SOURCED_PROFILE_SCRIPTS`, `dir_set` dedups, idempotency test); the ordering algorithm is correct (reverse iteration + `push_front`, layered test asserts inner/bin : outer/prepend : outer/bin); empty/sentinel handling is sound across all four shells; and both new bats tests genuinely fail without the fix.

Overall verdict: a clean, well-factored, thoroughly unit-tested extension of existing machinery that introduces no new injection class — the substantive gap is integration-test posture (bash-only, never attaches) for a PR titled "replay across attachments," so I recommend adding a real attach test and a tcsh case before merge, and resolving the three Important findings and the run-mode documentation gap.

## Domain Questions

Run-mode exclusion is undocumented: `flox_prepend_path` is inert in run mode because run mode does not source package `profile.d` scripts (per PR 4509 / issue 4514), yet the new docs describe the helper as preserving additions "as environments are activated, layered, and attached to" without noting this. Should the limitation be documented?

Layer-aware vs global placement semantics: per-layer placement is a real semantic choice — packages whose binaries must shadow tools globally regardless of layering order will not get that from this helper. Is per-layer the intended granularity for all use cases, or is a global-override escape hatch expected?

Profile.d script ordering vs registration: a `flox_prepend_path` call from a script that sorts before Flox's own path-setup scripts registers against an env whose path machinery has not yet run in that pass — worth confirming there is no hazard here.

Tie to issue 4514: the ordering-hazard question above becomes more pointed once the 4514 reserved-prefix convention is settled — confirming the interaction now would avoid a rework later.

---
*Via Forge (forge-review) • b368011e*

File: assets/environment-interpreter/common/activate.d/helpers.bash [hunk 1]
```diff
@@ -35,6 +35,29 @@ function source_profile_d {
...
+  local _env="${2:-${FLOX_ENV?}}"
+  case ":${_FLOX_ENV_PATH_PREPENDS:-}:" in
+    *":$_env=$_dir:"*) : ;; # already registered
+    *)
+      export _FLOX_ENV_PATH_PREPENDS="$_env=$_dir${_FLOX_ENV_PATH_PREPENDS:+:$_FLOX_ENV_PATH_PREPENDS}"
+      ;;
+  esac
+  # Take effect immediately in this shell; subsequent fix-paths
+  # invocations keep the dir ordered with its environment's layer.
+  PATH="$_dir:$PATH"
```
[the remark this answers is not in the document]
Comment (limeytexan) [reply assets/environment-interpreter/common/activate.d/helpers.bash]: Good catch — applied in 079e8ac0b, with the prepend moved inside the unregistered branch exactly as suggested. Command mode was indeed the flow where the duplicate would have been user-visible, so the test package now registers the same dir from two profile.d scripts and the in-place test additionally asserts that a command-mode activation sees exactly one PATH entry for the dir (that assertion fails against the previous unconditional prepend).

File: cli/tests/activate.bats [hunk 2]
```diff
@@ -2752,6 +2752,51 @@ EOF
...
+  rendered="$PROJECT_DIR/.flox/run/${NIX_SYSTEM}.${PROJECT_NAME}-dev"
+  run bash -c 'eval "$("$FLOX_BIN" activate)"; echo "$PATH"'
+  assert_success
+  assert_line --partial "$rendered/extra-bin:$rendered/bin:$rendered/sbin"
+}
+
+# When environments are layered, a registered prepend stays with the layer
+# of the environment that registered it: environments activated later still
+# take precedence over it.
+@test "profile.d: flox_prepend_path additions are replayed at their environment's layer" {
```
[the remark this answers is not in the document]
Comment (limeytexan) [reply cli/tests/activate.bats:2863]: Added in 32dbb8f0e. The new test keeps a command-mode activation running in the background (blocked on a fifo, per the idiom used by the other attach tests) and then attaches twice — once in command mode and once in-place — asserting the registered dir stays immediately ahead of the environment's `bin`/`sbin` in both. That covers the attach-side wiring end-to-end: the replayed `_FLOX_ENV_PATH_PREPENDS` from the recorded activation state feeding both the Rust-side export computation and the trailing `fix-paths` in the generated rc.

**Comment by limeytexan** [comment]
Thanks for the thorough review. Note it ran against `006f819`; two findings were already fixed by the time it posted, and the rest are now addressed in `0ced88561`. Item by item:

**Important findings:**

- **No end-to-end attach coverage** — addressed in `32dbb8f0e`: a background activation is held open on a fifo and both a command-mode and an in-place attach assert the registered dir stays ahead of the environment's `bin`/`sbin`.
- **Bash-only integration tests** — addressed: the in-place test is now parametrized across bash, fish, tcsh, and zsh. The tcsh case exercises the sentinel dance end-to-end and additionally asserts after deactivate that neither the sentinel nor a registration leaks into the shell.
- **Silent drop with the two-arg helper form** — addressed by dropping the optional `[<flox_env>]` argument entirely. It was speculative; registrations are now always keyed by `$FLOX_ENV`, which is byte-identical to the `FLOX_ENV_DIRS` entry by construction, so the exact-match hazard can no longer be reached.

**Minor findings:**

- **`:`/`=` in dir paths** — documented as a constraint in both the helper comment and `flox-activate(1)`.
- **Version-skew coupling** — verified there is no skew window: the rc statements embed `FLOX_ACTIVATIONS_BIN` (the generating binary's own path) and the interpreter scripts invoke `@flox_activations@`, substituted at interpreter build time — so the `--path-prepends` emitter and the `fix-paths` parser are always the same build in both flows.
- **Accumulate-vs-prune** — documented on `separate_prepends_list`: entries accumulate in the variable, filtering happens at replay, and deactivation's env-diff restoration bounds growth to the activation stack depth.
- **Deactivate unverified** — now verified two ways: decoding `_FLOX_HOOK_DIFF` confirms `_FLOX_ENV_PATH_PREPENDS` is tracked as `added`, and new bats assertions confirm the variable does not survive `flox deactivate` in bash and tcsh.
- **Unit test edge gaps** — added all three: a dir registered by two active envs (dedups to the outermost registering env's layer — it keeps the position it already held when the inner env layered on top, exactly as an inherited PATH entry would), a prepend equal to the env's own `bin` (dedups away), and dirs containing spaces (round-trip).
- **Unconditional re-prepend on guard miss** — this was fixed in `079e8ac0b` before the review posted; the prepend now lives inside the unregistered branch.
- **Serde inconsistency** — aligned with the `full_env` sibling (`skip_serializing_if`).
- **Quoting posture, O(env_dirs × prepends), tcsh divergence, threading duplication** — acknowledged as on-record observations; no change.

**Domain questions:**

- **Run-mode exclusion** — now documented in `flox-activate(1)`: `flox_prepend_path` has no effect in run mode because run mode does not source package-provided startup scripts (deferred design question tracked in #4514).
- **Layer-aware vs global placement** — per-layer is the intended semantic: a prepend is an extension of its environment's layer, not a global override, mirroring how the environment's own `bin` behaves. A global-override escape hatch is deliberately out of scope; if a real use case appears the entry format can grow an opt-in marker without breaking existing registrations.
- **Registration before Flox's own path scripts** — no hazard: registration only appends to the variable and to `PATH`; placement is computed at fix-time from `FLOX_ENV_DIRS`, which is established before profile.d sourcing begins in every mode, so a script that sorts ahead of `0100_*` still registers and replays correctly.
- **Tie to #4514** — agreed; I'll note `flox_prepend_path` in the #4514 discussion so the prefix-convention decision accounts for it.

File: assets/environment-interpreter/common/activate.d/helpers.bash [hunk 3]
```diff
@@ -35,6 +35,33 @@ function source_profile_d {
...
+#
+# Prepend <dir> to PATH and register it against <flox_env> (default:
+# $FLOX_ENV) in _FLOX_ENV_PATH_PREPENDS so that later invocations of
+# 'flox-activations fix-paths' replay the prepend at that environment's
+# position in the layered PATH rather than demoting it behind the
+# activated environments' bin directories. Intended for use by
+# etc/profile.d scripts.
+function flox_prepend_path {
+  local _dir="${1?}"
+  local _env="${2:-${FLOX_ENV?}}"
```
[the remark this answers is not in the document]
Comment (limeytexan) [reply assets/environment-interpreter/common/activate.d/helpers.bash]: Resolved in 0ced88561 by taking the first option: the `[<flox_env>]` parameter is dropped entirely until there's a caller for it. Registrations are now always keyed by `$FLOX_ENV`, which is byte-identical to its `FLOX_ENV_DIRS` entry by construction, so the trailing-slash/symlink mismatch can no longer be expressed. (The related no-diagnostic concern is also addressed — see the reply on the encoding thread: parsed-but-unmatched entries now emit a debug-level log, 55b22a9df.)

File: assets/environment-interpreter/common/activate.d/helpers.bash [hunk 4]
```diff
@@ -35,6 +35,33 @@ function source_profile_d {
...
+  case ":${_FLOX_ENV_PATH_PREPENDS:-}:" in
+    # Already registered: within an activation a registration is always
+    # accompanied by its PATH entry (set together below, or inherited
+    # along with the PATH that already contains the dir), so skip the
+    # prepend as well to keep repeated calls from duplicating it.
+    *":$_env=$_dir:"*) : ;;
+    *)
+      export _FLOX_ENV_PATH_PREPENDS="$_env=$_dir${_FLOX_ENV_PATH_PREPENDS:+:$_FLOX_ENV_PATH_PREPENDS}"
```
[the remark this answers is not in the document]
Comment (limeytexan) [reply assets/environment-interpreter/common/activate.d/helpers.bash:61]: Addressed across two commits: 0ced88561 documents the character constraint (`<dir>` must not contain `:` or `=`) in both the helper comment and `flox-activate(1)`, and removes the `[<flox_env>]` argument, closing that route entirely. For the remaining reachable case you identified — `$FLOX_ENV` itself containing `=` via the project path, which mangles the default one-arg form — 55b22a9df adds the suggested diagnostic: `prepend_dirs_to_pathlike_var` now emits a debug-level log for any parsed entry whose environment matches no `FLOX_ENV_DIRS` member, so the works-then-vanishes-on-attach failure is observable with `flox activate -vv` instead of silent. The helper comment also now notes the environment path is subject to the same constraint. Full escaping of the encoding feels like over-engineering while Nix store paths and sane project paths are `=`-free, but the entry format can grow escaping compatibly if it's ever warranted.

**Comment by limeytexan** [comment]
Thanks for the re-review and the hand-trace of `55b22a9`.

On the one non-blocking observation: agreed, and leaving the wording as-is per your recommendation — `split_once('=')` tolerating a `=` in the *dir* is an implementation detail of the current parser, and documenting the conservative constraint means we never have to promise that behavior if the entry format ever grows escaping.

Since the review note mentions the suite couldn't be run outside the dev shell, for the record the full matrix was run locally against `55b22a9df`: 108 `flox-activations` unit tests (including the 8 prepend-ordering tests), clippy clean, all 7 `flox_prepend_path` bats tests (four-shell in-place with bash/tcsh deactivate-leak assertions, command-mode idempotency, layering, genuine attach), plus the 77 profile, 51 attach, 13 activate-mode, and 41 deactivate suites — all passing.

With all four rounds resolved and the assessment at "Ready to merge", this is now ready for human review/merge.

**Comment by limeytexan** [comment]
Verified the item-by-item response against commits 079e8ac0b, 32dbb8f0e, 0ced88561, and 55b22a9df, plus the subsequent re-review rounds on this thread. All findings from this review are resolved. Item by item:

**[I] No end-to-end attach coverage — resolved** in 32dbb8f0e. The new test holds a background activation open on a fifo (the file's established idiom) and exercises both a command-mode and an in-place attach, asserting the registered dir stays immediately ahead of the environment's own `bin`/`sbin`. This is the first test to drive the `fixed_vars_to_export` attach path the title describes, and it converts the "across attachments" claim from threading-by-construction into demonstrated behavior.

**[I] Bash-only integration tests — resolved** in 0ced88561. The in-place test is now parametrized across bash, fish, tcsh, and zsh, and the tcsh case both executes the sentinel dance end-to-end and asserts after deactivate that neither the sentinel nor a registration leaks. That goes beyond what the finding asked for — the sentinel's non-leak property is now regression-tested rather than argued.

**[I] Silent drop with the two-arg helper form — resolved by the strongest available means:** the speculative `[flox_env]` argument is gone entirely. Registrations are always keyed by `$FLOX_ENV`, which is byte-identical to the `FLOX_ENV_DIRS` entry by construction, so the exact-match hazard is unreachable rather than merely documented. The later `debug!` diagnostic in 55b22a9df additionally surfaces any residual parse-and-drop case (for example an environment path mangled by the encoding constraint), which closes the observability gap the finding pointed at.

**Minor findings** — all confirmed: the `:`/`=` constraint is documented in both the helper comment and `flox-activate(1)` (and keeping the conservative wording rather than promising `split_once` leniency is the right call — it preserves freedom to add escaping later); accumulate-vs-prune is documented on `separate_prepends_list` with deactivation's env-diff restoration bounding growth; deactivate is now verified two ways (hook-diff decoding plus bats assertions in bash and tcsh); the three unit-test edge cases were added, and the two-envs-same-dir semantics (keeps the position it held when the inner env layered on top, like any inherited PATH entry) is the coherent choice; the unconditional re-prepend was fixed in 079e8ac0b, which also makes command mode — the one flow with no trailing fix-paths — genuinely idempotent; serde is aligned with `full_env`.

**Version-skew — withdrawn on verification.** I confirmed on the branch that the interpreter hardcodes `@flox_activations@`, substituted at interpreter build time (activate, line 29), and the generated rc statements embed the generating binary's own `FLOX_ACTIVATIONS_BIN` path. The `--path-prepends` emitter and the `fix-paths` parser are therefore always the same build in both flows; there is no skew window and no action needed.

**Domain questions** — all settled: the run-mode exclusion is now documented in `flox-activate(1)` with the design question tracked in issue 4514; per-layer placement confirmed as the intended semantic with a plausible extension path (an opt-in marker in the entry format) if a global-override use case ever materializes; and the ordering-hazard rebuttal is sound — registration only appends, while placement is computed at fix time from `FLOX_ENV_DIRS`, which every mode establishes before profile.d sourcing begins, so a script sorting ahead of the interpreter's own path scripts still registers and replays correctly. Noting `flox_prepend_path` in the issue 4514 prefix-convention discussion is the right follow-through.

For the record, the locally-run full matrix reported above (108 unit tests, clippy clean, all 7 `flox_prepend_path` bats tests, plus the profile/attach/activate-mode/deactivate suites) addresses the only caveat left standing in the automated re-reviews, which could not run the suite outside the dev shell.

Nothing further from this review — all four rounds are resolved and regression-tested where behavioral. Clear for human review and merge.

---
*Via Forge (forge-review) • b368011e*

File: cli/flox-activations/src/gen_rc/fish.rs [hunk 5]
```diff
@@ -175,7 +175,7 @@ pub fn generate_fish_profile_commands(
             );
 
             stmts.push(format!(
-                r#"{} fix-paths --shell {} --env-dirs "$FLOX_ENV_DIRS" --path "$PATH" --manpath "$MANPATH" | source;"#,
+                r#"{} fix-paths --shell {} --env-dirs "$FLOX_ENV_DIRS" --path "$PATH" --manpath "$MANPATH" --path-prepends "$_FLOX_ENV_PATH_PREPENDS" | source;"#,
```
[the remark this answers is not in the document]
Comment (limeytexan) [reply cli/flox-activations/src/gen_rc/fish.rs:178]: Both points addressed in 640b0987f:

1. **Latent naming constraint** — documented on the `FLOX_ENV_PATH_PREPENDS_VAR` definition: the name must not end in `PATH` because fish auto-splits `*PATH` variables on import and re-joins on expansion, which would corrupt the multi-entry value in transit.
2. **Multi-entry transport coverage** — rather than adding a separate layered fish/tcsh test, the test package now registers a *second* dir (`0902_` script → `extra-bin2`), so `_FLOX_ENV_PATH_PREPENDS` holds a genuine colon-delimited two-entry value in every test. Each of the four per-shell in-place tests therefore transports the multi-entry value through that shell's own variable handling — fish quoted expansion, tcsh `:q`, zsh `${…:-}` — and asserts the full `extra-bin2:extra-bin:bin:sbin` layer ordering (most recent registration foremost). The layering, attach, and command-mode idempotency assertions were strengthened to the two-dir chains as well.

All 108 unit tests and 7 integration tests pass locally with the change.
